casbin/enforcer_cached.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+* Copyright 2020 The casbin Authors. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#pragma once
+
+#include "pch.h"
+
+#include "./enforcer_cached.h"
+using namespace std;
+
+CachedEnforcer::CachedEnforcer(CachedEnforcer& ce){
+    this->e = move(ce.e);
+    this->enableCache = ce.enableCache;
+    this->m = ce.m;
+}
+
+CachedEnforcer::CachedEnforcer(unique_ptr<Enforcer>& e) {
+  this->e = move(e);
+  this->enableCache = false;
+}
+
+void CachedEnforcer::EnableCache(const bool& enableCache) {
+  this->enableCache = enableCache;
+}
+
+bool CachedEnforcer::Enforce(const vector<string>& rval) {
+  if (!enableCache) {
+    return e->Enforce(rval);
+  }
+
+  string key;
+  for (auto r : rval) {
+    key += r;
+    key += "$$";
+  }
+
+  pair<bool, bool> res_ok = getCachedResult(key);
+
+  if (res_ok.second) {
+    return res_ok.first;
+  }
+
+  bool res = e->Enforce(rval);
+  setCachedResult(key, res);
+  return res;
+}
+
+pair<bool, bool> CachedEnforcer::getCachedResult(const string& key) {
+  locker.lock();
+  bool ok = m.count(key);
+  if (!ok) {
+    locker.unlock();
+    return pair<bool, bool>(false, false);
+  }
+
+  pair<bool, bool> res_ok(m[key], ok);
+  locker.unlock();
+  return res_ok;
+}
+
+void CachedEnforcer::setCachedResult(const string& key, const bool& res) {
```

Comment:
How about using a concurrent hashmap here, so no need to implement a lock?

Reply:
I think use of standard mutex locks is more efficient than importing or implementing concurrent hash-maps.

Reply:
We don't have a concurrent hashmap in cpp.